### DISCUSSION TITLE
Lifecycle of events and callbacks

### DISCUSSION THREAD

**Discussion by DavMoo** [orestbida/cookieconsent#414] [General]
Dear Orest,

I am facing the following situation:

When i accept the cookie banner i trigger a page reload within the `onFirstConsent` callback.
Now when the page got reloaded i **first** get a test console output from my test script which got allowed via cookie banner.
The **second** console output i get is coming from `onConsent` callback.

This seems to be not correct. At least i expect the `onConsent` callback being fired before any script-tag is getting allowed.

Do you have an explanation for that?

Thanks

David

**Comment by orestbida** [comment]
Hi @DavMoo,

The `onFirstConsent` and `onConsent` callbacks/events are last in the "execution chain", and this is just a choice I made. When these 2 callbacks/events are fired, I'd like to be able to see the "whole final picture", with the most up-to-date version of the DOM.

Execution chain:

1. `onModalReady`
2. `onModalShow`

User makes a choice:

3. `onModalHide`
4. set main cookie
5. run autoclear function (if needed)
6. run internally configured services (if any)
7. enable `script` tags — categories and/or services — sequentally (if any)
8. `onFirstConsent`
9. `onConsent`

Next page load (if consent is still valid):
1. run internally configured services (if any)
2. enable `script` tags — categories and/or services — sequentally (if any)
3. `onConsent`

**Reply by orestbida** [reply]
I might create a dedicated section in the docs., so that it's clear to everyone.

**Reply by DavMoo** [reply]
Hey Orest,
Thank you for the clarification. That is helpful.
A dedicated section for the lifecycles would be nice to see in the docs.

Cheers.

**Comment by DavMoo** [comment]
Hey again.
I am struggling with the following issue.

I have both inline script tags and tags fired via the tag manager.

After user gives consent i need to reload the page to make full consent object available in tag manager after page got reloaded.

But as soon as the user clicks "accept all" the inline script tags fires before the page reload.
And a second time after the page reload (which is actually correct).

I need a way to prevent the inline scripts from being fired after the initial consent but only after the reload.

Do you have any idea for this of is it somehow covered in your code?

Thanks

David

**Reply by orestbida** [reply]
There is no ideal solution for this scenario, but there are still ways to achieve it.

You can use the `setCookieData` to set a specific value if the category is enabled and reload the page. The value can be retrieved via `getCookie('data')`;

```javascript
// cookieconsent-init.js
const reloadPage = () => {
    if(!CookieConsent.acceptedCategory('analytics'))
        return;

    CookieConsent.setCookieData({
        value: true
    });
    window.location.reload();
};

CookieConsent.run({
    onFirstConsent: () => {
        reloadPage();
    },

    onChange: ({changedCategories}) => {
        changedCategories.includes('analytics') && reloadPage();
    },
});
```
```html
<script type="text/plain" data-category="analytics">
    if(CookieConsent.getCookie('data') === true){
        console.log("enabled")
    }
</script>
```

You could also use `localStorage` instead of the cookie's `data` field, if the API is not reachable.
